validation: set BLOCK_FAILED_CHILD correctly

stratospher opened the pull request:
This PR addresses 3 issues related to how `BLOCK_FAILED_CHILD` is set:
1. In `InvalidateBlock()`
- Previously, `BLOCK_FAILED_CHILD` was not being set when it should have been.
- This was due to an incorrect traversal condition, which is fixed in this PR.

2. In `SetBlockFailure()`
- `BLOCK_FAILED_VALID` is now cleared before setting `BLOCK_FAILED_CHILD`.

3. In `InvalidateBlock()`
- if block is already marked as `BLOCK_FAILED_CHILD`, don't mark it as `BLOCK_FAILED_VALID` again.

Also adds a unit test to check `BLOCK_FAILED_VALID` and `BLOCK_FAILED_CHILD` status in `InvalidateBlock()`.

<details>
<summary><h3>looking for feedback on an alternate approach</h3></summary>
<br>

An alternate approach could be removing `BLOCK_FAILED_CHILD` since even though we have a distinction between
`BLOCK_FAILED_VALID` and `BLOCK_FAILED_CHILD` in the codebase, we don't use it for anything. Whenever we check for BlockStatus, we use `BLOCK_FAILED_MASK` which encompasses both of them. See  similar discussion in https://github.com/bitcoin/bitcoin/pull/16856.

I have a branch with this approach in https://github.com/stratospher/bitcoin/commits/2025_02_remove_block_failed_child/.
Compared to the version in #16856, it also resets `BLOCK_FAILED_CHILD` already on disk to `BLOCK_FAILED_VALID` when loading from disk so that we won't be in a dirty state in a no-`BLOCK_FAILED_CHILD`-world.

I'm not sure if it's a good idea to remove `BLOCK_FAILED_CHILD` though. would be curious to hear what others think of this approach.

thanks @ mzumsande for helpful discussion regarding this PR!
</details>

stratospher committed:
test: check BlockStatus when InvalidateBlock is used

when a block is invalidated using InvalidateBlock, check that:
1. it's status is BLOCK_FAILED_VALID
2. it's children's status is BLOCK_FAILED_CHILD
   and not BLOCK_FAILED_VALID
3. it's ancestors are valid

stratospher committed:
validation: fix traversal condition to mark BLOCK_FAILED_CHILD

this block of code is not reached on master since other than
initialisation, all other iterations have invalid_walk_tip
and to_mark_failed pointers in some form of this layout
where 1, 2, 3 and 4 are block heights.

	invalid_walk_tip
	  ↓
1 <- 2 <- 3 <- 4
	       ↑
	      to_mark_failed

fix it so that blocks are correctly marked as BLOCK_FAILED_CHILD
if it's a descendant of BLOCK_FAILED_VALID block.

stratospher committed:
validation: correctly update BlockStatus for invalid block descendants

invalid_block ----------> block_index

- before this commit, only if block_index is not invalid, it will mark
  block_index as BLOCK_FAILED_CHILD
- it's possible that block_index encountered is invalid and was marked
  as BLOCK_FAILED_VALID previously
- in this case, correctly update BlockStatus of block_index by
  clearing BLOCK_FAILED_VALID and then setting it to BLOCK_FAILED_CHILD

mzumsande reviewed (commented):
Concept ACK

This is a straightforward bugfix.

In general, I'm in favor of removing the `BLOCK_FAILED_CHILD` / `BLOCK_FAILED_VALID` distinction as a follow-up, as suggested in https://github.com/bitcoin/bitcoin/pull/16856#issuecomment-565506585. It's been introduced in 2012 (#1677) and (as far as I know) has never been used for anything. I also can't think of a legitimate use - block indexes that are marked as invalid are invalid indefinitely (outside of the `reconsiderblock` rpc, but that removes both flags from all descendants of a reconsidered block so it doesn't need the distinction either.)  It just adds unnecessary complexity in my opinion.

TheBlueMatt committed:
validation: clarify final |= BLOCK_FAILED_VALID in InvalidateBlock

This has no functional affect, as the any CBlockIndex*s which
to_mark_failed is set to will already have been marked failed.

Also prevents a situation where block already marked as
BLOCK_FAILED_CHILD is again unconditionally marked as
BLOCK_FAILED_VALID in the final |= BLOCK_FAILED_VALID.

mzumsande reviewed (commented):
Code Review ACK 17718660b8c95d1c12124ba2f38baf286a3bddf2

TheCharlatan commented on src/test/blockchain_tests.cpp:
```diff
@@ -117,4 +117,36 @@ BOOST_AUTO_TEST_CASE(num_chain_tx_max)
+    WITH_LOCK(::cs_main, assert(tip_to_invalidate->nStatus & BLOCK_FAILED_VALID));
+    WITH_LOCK(::cs_main, assert((tip_to_invalidate->nStatus & BLOCK_FAILED_CHILD) == 0));
+
+    // check all ancestors of block are validated up to BLOCK_VALID_TRANSACTIONS and are not invalid
```
Nit: Should this say "Check all ancestors of the invalidated block are validated up to..."?

sedited reviewed (approved):
ACK 17718660b8c95d1c12124ba2f38baf286a3bddf2

This looks fine, but I think the commit messages could all be a bit clearer.

sedited reviewed (approved):
Re-ACK 3c3548a70eedb8dcf6a4a8d605a4a12e814c7cac

mzumsande reviewed (commented):
re-ACK 3c3548a70eedb8dcf6a4a8d605a4a12e814c7cac

naiyoma reviewed (commented):
Concept ACK
Tested  [https://github.com/bitcoin/bitcoin/pull/31835/commits/c99667583dd9b57612edf4c04611cd4857250600](https://github.com/bitcoin/bitcoin/pull/31835/commits/c99667583dd9b57612edf4c04611cd4857250600) on regtest using 5 blocks and invalidated block 3.
Before the traversal fix:
invalidated block and all its children were marked as BLOCK_FAILED_VALID.

                    ↓[BLOCK_FAILED_VALID]
    1<-----2<-------3<------4<------5
                            ↑       ↑
                            ↑       [BLOCK_FAILED_VALID]
                            [BLOCK_FAILED_VALID]

After the fix, the invalidated block is now marked as BLOCK_FAILED_VALID, and all its children are marked as BLOCK_FAILED_CHILD.
     
                            ↓invalid_walk_tip[BLOCK_FAILED_VALID]
    1<-----2<-------3<------4<------5
                                    ↑to_mark_failed[BLOCK_FAILED_CHILD] 
                                    
    
                    ↓invalid_walk_tip[BLOCK_FAILED_VALID]
    1<-----2<-------3<------4<------5
                            ↑to_mark_failed[BLOCK_FAILED_CHILD]
                            
        
                    ↓[BLOCK_FAILED_VALID]
    1<-----2<-------3<------4<------5
                            ↑       ↑[BLOCK_FAILED_CHILD]
                            ↑[BLOCK_FAILED_CHILD]

naiyoma commented:
> I'm not sure if it's a good idea to remove `BLOCK_FAILED_CHILD` though. would be curious to hear what others think of this approach.

IMO, I think removing BLOCK_FAILED_CHILD is a better approach, I agree with this comment [https://github.com/bitcoin/bitcoin/pull/16856#issuecomment-687724324](https://github.com/bitcoin/bitcoin/pull/16856#issuecomment-687724324),  it seems  simpler and helps avoid potential hidden bugs
